### PR TITLE
Prototype new proposed API for histogramming and friends

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This is a draft, just want to hear some initial opinions.

Suggestion for new API of `histogram`:

- Use keyword arg, accepting integer, bin size (as scalar variable), or actual edges.
- Alternatively, accept a `dict` (e.g., when coord name is not a valid variable name).

The plan is to support the same in other cases, such as the `edges` argument of `sc.bin`. I am not sure a keyword syntax makes sense in that case, since we also have the `groups` and `erase` arguments... unless someone can think of a clean way to distinguish?

`sc.groupby` is another candidate for this updated API. Here the `bins` keyword argument should change, but I am not sure yet how exactly.

### CODE REVIEW

**Comment by jl-wynen** [comment]
> I am not sure a keyword syntax makes sense in that case,

Without having read the code yet, I am not sure that syntax makes sense here either. It limits us in adding new arguments to the function. Are we sure that we never need to?

File: src/scipp/core/bins.py [hunk 1]
```diff
@@ -394,7 +424,12 @@ def histogram(x: Union[_cpp.DataArray, _cpp.Dataset], *,
     scipp.bin:
         For binning data.
     """
-    return _call_cpp_func(_cpp.histogram, x, bins)
+    if edges is not None:
+        kwargs = dict(**edges, **kwargs)
```
Comment (jl-wynen) [inline src/scipp/core/bins.py:428]: `kwargs` overrides `edges`. Is this intentional? Or should we raise on clashes?

**Comment by SimonHeybrock** [comment]
> > I am not sure a keyword syntax makes sense in that case,
> 
> Without having read the code yet, I am not sure that syntax makes sense here either. It limits us in adding new arguments to the function. Are we sure that we never need to?

I don't think we can ever be sure. We may consider https://numpy.org/doc/stable/reference/generated/numpy.histogram.html as a guide, which has additional `weights` and `density` arguments. `weight` would typically be given by our variances, so that does not apply. Instead of adding a `density` argument, we would likely have to make a new named function? Or limit the keyword args that are available to the user?

So in summary:

- This does not necessarily prevent us from adding keyword args, it would simply make it a breaking change.
- We have to find a good balance between usability for the 95% of applications, and keeping all options open. I am not sure where the balance is.

File: src/scipp/core/bins.py [hunk 2]
```diff
@@ -378,8 +379,37 @@ def _groupby_bins(obj):
...
+    bound.value = np.nextafter(bound.value, np.inf)
+    return bound
+
+
+def _parse_coords_arg(x, name, arg):
+    coord = _get_coord(x, name)
+    if isinstance(arg, int):
+        coord = linspace(name, coord.min(), _upper_bound(coord), num=arg + 1)
+    elif isinstance(arg, Variable):
+        if arg.ndim == 0:
```
Comment (jl-wynen) [inline src/scipp/core/bins.py:399]: Is this a relevant case in practice? When does one have a known bin size but arbitrary boundaries?

Also, What about a scalar variable with `unit = None`? Would this be like passing an `int`?

File: src/scipp/core/bins.py (hunk 1 above)
Comment (SimonHeybrock) [reply src/scipp/core/bins.py:428]: This syntax *does* raise?

File: src/scipp/core/bins.py (hunk 2 above)
Comment (SimonHeybrock) [reply src/scipp/core/bins.py:399]: I think it is relevant. Someone may not care about the bounds, but knows their instrument resolution and want "0.1 Angstrom" bins.

Good point, need to avoid calling `to_unit` when `unit = None`. No, it would *not* be like passing and `int`.

**Comment by jl-wynen** [comment]
Given bootstrapping and other error estimation techniques, `resample` is a poor name. Histogramming and binning are ways to estimate distributions, not sample from them. Not sure grouping falls into the same category. So I am unsure what a good name would be.

**Comment by SimonHeybrock** [comment]
I don't like the two-step syntax anyway :-D

**Comment by SimonHeybrock** [comment]
I think ultimately we need to ensure things are consistent. `sc.histogram` is essentially a "direct" path replacing `sc.bin(da, ...).bins.sum()`, so we may eventually want to move to a place where both support all options? For example, currently `sc.histogram` cannot to what `sc.bin(da, ...).bins.mean()` does, let alone `nansum` or `nanmean`.

We could consider simpler dedicated functions:

```python
sc.histogram(x=100)
sc.histogram(x=100, y=100)  # future support
sc.weighted_histogram(x=100)  # unless we want to have reserved keyword args
sc.rebin(x=100)

# Replace sc.bin by 3 higher-level functions (rename sc.bin to sc.modify_bins or something specific)
sc.bin(x=100)
sc.group('pixel')
sc.group(pixel=selected_pixels)
sc.bins.concat('x')
```

But the issue of `histogram` being tied to `sum` remains. Maybe a 2-step solution *is* required?
```python
da.hist(x=100).sum()
da.hist('year').mean()
```

Now this looks almost like `groupby`, except that `sc.histogram` uses a different algorithm that is faster, with different capabilities and limitations in terms of shapes, etc.

**Comment by SimonHeybrock** [comment]
> > But the issue of histogram being tied to sum remains. Maybe a 2-step solution is required?
> 
> Isn't this what the `density` argument in numpy does?

Not quite, that also normalizes? Also, for me the docs are unclear: Does it just normalize, or compute the mean in each bin and then normalize?

**Comment by jl-wynen** [comment]
I think it is `sum(events) / bin_width / n_events_total`. So something that is entirely missing from scipp.

**Comment by SimonHeybrock** [comment]
> I think it is `sum(events) / bin_width / n_events_total`. So something that is entirely missing from scipp.

That can easily be done in a second step then, no need to add confusing keyword arguments. ;)

**Comment by SimonHeybrock** [comment]
I have implemented what we discussed on this branch, but I"ll open a new PR to avoid partially outdated/unrelated discussions in this long thread.
